# Data Ingestion

In [222]:
import gdown

# Downloading dataset from drive

In [223]:
dataset_url = "https://drive.google.com/uc?export=download&id=1igwnF7UazJnk4b-Vebh1uH25nWGjWowa"
filename_train = "crop_health_train_data.csv"

In [224]:
gdown.download(dataset_url, filename_train)

Downloading...
From: https://drive.google.com/uc?export=download&id=1igwnF7UazJnk4b-Vebh1uH25nWGjWowa
To: G:\DataScience and Ml using Python\week8\Day4\crop_health_train_data.csv
100%|█████████████████████████████████████████████████████████████████████████████| 2.91M/2.91M [00:00<00:00, 3.46MB/s]


'crop_health_train_data.csv'

# Importing dataset

In [225]:
train_df = pd.read_csv(filename_train)
train_df.head()

ID  Estimated_Insects_Count  Crop_Type  Soil_Type  \
0  F00000001                      188          1          0   
1  F00000003                      209          1          0   
2  F00000004                      257          1          0   
3  F00000005                      257          1          1   
4  F00000006                      342          1          0   

   Pesticide_Use_Category  Number_Doses_Week  Number_Weeks_Used  \
0                       1                  0                0.0   
1                       1                  0                0.0   
2                       1                  0                0.0   
3                       1                  0                0.0   
4                       1                  0                0.0   

   Number_Weeks_Quit  Season  Crop_Damage  
0                  0       1            0  
1                  0       2            1  
2                  0       2            1  
3                  0       2            1  
4                  0       2            1

## About dataset

## Importingimportant libraries for this project

In [226]:
import numpy as np   # for numerical operation
import pandas as pd    # for data analysis
import matplotlib.pyplot as plt    # data visualization
import seaborn as sns    # data visualization

# Data Profiling and Inspection

#### Dtypes

In [227]:
# checking the type of data in dataset
train_df.dtypes

ID                          object
Estimated_Insects_Count      int64
Crop_Type                    int64
Soil_Type                    int64
Pesticide_Use_Category       int64
Number_Doses_Week            int64
Number_Weeks_Used          float64
Number_Weeks_Quit            int64
Season                       int64
Crop_Damage                  int64
dtype: object

* Here data type in this dataset are object, integer and floalt

#### Descritive statistics 

In [228]:
train_df.describe()   # for numerical data

Estimated_Insects_Count     Crop_Type     Soil_Type  \
count             88858.000000  88858.000000  88858.000000   
mean               1399.012210      0.284375      0.458417   
std                 849.048781      0.451119      0.498271   
min                 150.000000      0.000000      0.000000   
25%                 731.000000      0.000000      0.000000   
50%                1212.000000      0.000000      0.000000   
75%                1898.000000      1.000000      1.000000   
max                4097.000000      1.000000      1.000000   

       Pesticide_Use_Category  Number_Doses_Week  Number_Weeks_Used  \
count            88858.000000       88858.000000       79858.000000   
mean                 2.264186          25.849952          28.623970   
std                  0.461772          15.554428          12.391881   
min                  1.000000           0.000000           0.000000   
25%                  2.000000          15.000000          20.000000   
50%                  2.000000          20.000000          28.000000   
75%                  3.000000          40.000000          37.000000   
max                  3.000000          95.000000          67.000000   

       Number_Weeks_Quit        Season   Crop_Damage  
count       88858.000000  88858.000000  88858.000000  
mean            9.589986      1.896959      0.190562  
std             9.900631      0.701322      0.454215  
min             0.000000      1.000000      0.000000  
25%             0.000000      1.000000      0.000000  
50%             7.000000      2.000000      0.000000  
75%            16.000000      2.000000      0.000000  
max            50.000000      3.000000      2.000000

##### From above descritive statistics 
1. Estimated_Insects_Count :
    * Mean of Estimated_Insects_Count is 1399.01
    * Standard Deviation is 849.048781
    * Minimum value in Estimated_Insects_Count is 150
    * Maximum value in Estimated_Insects_Count is 4097
    * 731.000000 data falls under the 25% 
    * 1898.000000 data falls under the 75%
2. Number_Doses_Week :
    * Mean of Number_Doses_Week is 25.849952
    * Standard Deviation is 15.554428
    * Minimum value in Number_Doses_Week is 0
    * Maximum value in Number_Doses_Week is 95
    * 15 data falls under the 25% 
    * 40 data falls under the 75%
3. As you can see, almost all the values of Crop_Type, Soil_Type seems to be either a "0" or a "1". So it is better to treat it as a categorical value rather than a numerical value. 
4. As you can see, almost all the values of Pesticide_Use_Category, Season seems to be either a "1" or a "2" or a "3". So it is better to treat it as a categorical value rather than a numerical value. 

#### Shape of dataset

In [229]:
train_df.shape

(88858, 10)

* Shape of data is (88858, 10) which mean total number of rows is 88858 and columns is 10

#### Overall information about dataset

In [230]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88858 entries, 0 to 88857
Data columns (total 10 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ID                       88858 non-null  object 
 1   Estimated_Insects_Count  88858 non-null  int64  
 2   Crop_Type                88858 non-null  int64  
 3   Soil_Type                88858 non-null  int64  
 4   Pesticide_Use_Category   88858 non-null  int64  
 5   Number_Doses_Week        88858 non-null  int64  
 6   Number_Weeks_Used        79858 non-null  float64
 7   Number_Weeks_Quit        88858 non-null  int64  
 8   Season                   88858 non-null  int64  
 9   Crop_Damage              88858 non-null  int64  
dtypes: float64(1), int64(8), object(1)
memory usage: 6.8+ MB


##### <b>From above overall information </b>
* total number of columns in this dataset is 10
* total number of entries(rows) in this dataset is 88858
* Number_Weeks_Used has 79858 non null data remianing are null

#### Checking missing data
* using .isna() methods

In [231]:
train_df.isna()

ID  Estimated_Insects_Count  Crop_Type  Soil_Type  \
0      False                    False      False      False   
1      False                    False      False      False   
2      False                    False      False      False   
3      False                    False      False      False   
4      False                    False      False      False   
...      ...                      ...        ...        ...   
88853  False                    False      False      False   
88854  False                    False      False      False   
88855  False                    False      False      False   
88856  False                    False      False      False   
88857  False                    False      False      False   

       Pesticide_Use_Category  Number_Doses_Week  Number_Weeks_Used  \
0                       False              False              False   
1                       False              False              False   
2                       False              False              False   
3                       False              False              False   
4                       False              False              False   
...                       ...                ...                ...   
88853                   False              False              False   
88854                   False              False              False   
88855                   False              False              False   
88856                   False              False              False   
88857                   False              False              False   

       Number_Weeks_Quit  Season  Crop_Damage  
0                  False   False        False  
1                  False   False        False  
2                  False   False        False  
3                  False   False        False  
4                  False   False        False  
...                  ...     ...          ...  
88853              False   False        False  
88854              False   False        False  
88855              False   False        False  
88856              False   False        False  
88857              False   False        False  

[88858 rows x 10 columns]

#### Get total missing values in each column 

In [232]:
train_df.isna().sum()

ID                            0
Estimated_Insects_Count       0
Crop_Type                     0
Soil_Type                     0
Pesticide_Use_Category        0
Number_Doses_Week             0
Number_Weeks_Used          9000
Number_Weeks_Quit             0
Season                        0
Crop_Damage                   0
dtype: int64

* From above  observation Number_Weeks_Used  has 9000 missing data

#### Missing data of Number_Weeks_Used in percentage

In [233]:
train_df.Number_Weeks_Used.isna().sum() / len(train_df.Number_Weeks_Used) * 100

10.128519660582052

* In this dataset 10% of data is missing in Number_Weeks_Used columns        

#### Visualizing missing data

In [ ]:
plt.figure(figsize=(10,10))
plt.title("Visualizing Missing Values")
sns.heatmap(train_df.isna().transpose(),cmap="YlGnBu",  # Set a yellow -> green -> blue color gradient
    cbar_kws={"label": "Missing Data"})
plt.show()

* From above observation the missing data has no any pattern so data are randomly missing.

#### Exploratory Data Analysis
* EDA is an approach of analyzing data sets to summarize their main characteristics

### Features Analysis

##### <b>Univariate Analysis</b>

In [ ]:
train_df.head(3)

#### Categorical Data Analysis

<b>1.</b> As you can see, almost all the values of Crop_Type, Soil_Type seems to be either a "0" or a "1". So it is better to treat it as a categorical value rather than a numerical value. Let's perform categorical analysis.

#### Crop_Type

In [ ]:
plt.figure(figsize=(8,8))
plt.title("Distribution of Soil_Type")
sns.countplot(x="Crop_Type",data=train_df, hue="Crop_Type", palette = "Set1")
plt.show()

##### <b>From above observation 
* Crop_Type "0" has around 65000 data entries in this dataset
* Crop_Type "1" has around 25000 data entries in this dataset
* Crop_Type data are imbalanced data.

#### Soil_Type

In [ ]:
soil_count = train_df.Soil_Type.value_counts()
plt.figure(figsize=(8,8))
plt.title("Distribution of Soil_Type")
sns.barplot(x=soil_count.index, y=soil_count.values, hue=soil_count.index, palette = "Set1")
plt.show()

##### <b>From above observation 
* Soil_Type "0" has around 49000 data entries in this dataset
* Soil_Type "1" has around 40000 data entries in this dataset
* Soil_Type data seems to be balanced data

<b>2.</b> As you can see, almost all the values of Pesticide_Use_Category, Season seems to be either a "1" or a "2" or a "3". So it is better to treat it as a categorical value rather than a numerical value. Let's perform categorical analysis.

#### Pesticide_Use_Category

In [ ]:
Pesticide_Use_Category = train_df.Pesticide_Use_Category.value_counts()
plt.figure(figsize=(8,8))
plt.title("Pesticide_Use_Category")
plt.pie(x=Pesticide_Use_Category.values, labels=Pesticide_Use_Category.index, autopct="%.3f%%", explode=[0.05,0,0.0], colors=["violet","pink","Cyan" ])
plt.legend()
plt.show()

#### From above observation 
* Pesticide_Use_Category "2" has highest percentage with 71.698% data in this dataset
* Pesticide_Use_Category "1" has lowest percentage with 0.942% data in this dataset
* Pesticide_Use_Category "2" has 27.361% data in this dataset

#### Season

In [ ]:
season_counts = train_df.Season.value_counts()
plt.title("Distribution of Season")
ax = season_counts.plot.pie(figsize=(8,8), explode=[0,0,0.05], autopct="%.3f%%",colors=["violet","pink","Cyan"])
plt.legend()

#### From above observation 
* Season "2" has highest percentage with 49.754% data in this dataset
* Season "3" has lowest percentage with 19.971% data in this dataset
* Season "1" has 30.275% data in this dataset

#### Numerical data analysis

##### Estimated_Insects_Count

In [ ]:
sns.displot(data=train_df.Estimated_Insects_Count, kde=True, height=10, color="purple")
plt.grid()
plt.show()

#### From above observation 
* Distribution of Estimated_Insects_Count seems to be right tail, which means mean > median

#### Number_Doses_Week

In [ ]:
sns.displot(data=train_df.Number_Doses_Week, kde=True, height=10, color="purple")
plt.show()

#### From above observation 
* Distribution of Number_Doses_Week also seems to be right tail, which means mean > median
* Data are distribute between 5 to 60

#### Number_Weeks_Used

In [ ]:
sns.displot(data=train_df.Number_Weeks_Used, kde=True, height=10,  color="purple")
plt.show()

#### From above observation 
* Distribution of Number_Weeks_Used seems to be normally distribution but slightly right skewed

#### Number_Weeks_Quit

In [ ]:
sns.displot(data=train_df.Number_Weeks_Quit, kde=True, height=10,  color="purple")
plt.show()

#### From above observation 
* Distribution of Number_Weeks_Quit seems to be right skewed which means mean > median
* Data are distributed between 0 to 40

#### Data Skewness

In [ ]:
categorical_features = ["Crop_Type","Soil_Type","Pesticide_Use_Category","Season"]
nemurical_features =["Estimated_Insects_Count","Number_Doses_Week","Number_Weeks_Used","Number_Weeks_Quit"]

In [ ]:
train_df[nemurical_features].skew()

# Target Variable Analysis: Crop_Damage

In [ ]:
c_d = train_df.Crop_Damage.value_counts()
sns.barplot(x=c_d.index, y=c_d.values, hue=c_d.index)
plt.legend()

In [ ]:

sns.barplot(x=soil_count.index, y=soil_count.values, hue=soil_count.index, palette = "Set1")
